### PR TITLE
Pod ssh and logs changes

### PR DESCRIPTION
**Pull request by rakol-ms**
Signed-off-by: Rajendra Kolli <rakol@microsoft.com>

### CODE REVIEW

**Comment by rakol-ms** [comment]
Pending tasks
- [ ] Double tab to get possible commands in bash
- [ ] Keyup for previous commands
- [ ] Support for pasting from clipboard
- [ ] Handle service accounts for RBAC clusters.
- [ ] Cleanup and consolidation of changes

File: package.json [hunk 1]
```diff
@@ -33,11 +34,14 @@
     "_bundles"
   ],
   "dependencies": {
-    "@kubernetes/client-node": "~0.7.1",
+    "@kubernetes/client-node": "~0.7.2",
+    "@types/node": "^10.12.18",
     "azure-devops-ui": "~1.144.1",
+    "isomorphic-ws": "^4.0.1",
```
Comment (madhuv-msft) [inline package.json]: can we live without this?

File: package.json [hunk 2]
```diff
@@ -33,11 +34,14 @@
     "_bundles"
   ],
   "dependencies": {
-    "@kubernetes/client-node": "~0.7.1",
+    "@kubernetes/client-node": "~0.7.2",
+    "@types/node": "^10.12.18",
```
Comment (madhuv-msft) [inline package.json]: is this a dev or product dependency?

File: src/Contracts/KubeServiceBase.ts [hunk 3]
```diff
@@ -34,11 +36,19 @@ export abstract class KubeServiceBase implements IKubeService {
...
-        return this.fetch(KubeResourceType.DaemonSets)
+        return this.fetch(KubeResourceType.DaemonSets);
     }
 
     getStatefulSets() : Promise<K8sTypes.V1StatefulSetList> {
-        return this.fetch(KubeResourceType.StatefulSets)
+        return this.fetch(KubeResourceType.StatefulSets);
+    }
+
+    getPodLogs(podName:string): Promise<string> {
```
Comment (madhuv-msft) [inline src/Contracts/KubeServiceBase.ts]: UTs?

File: src/Contracts/TerminalService.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
+import WebSocket = require('isomorphic-ws');
```
Comment (madhuv-msft) [inline src/Contracts/TerminalService.ts]: import WebSocket = require('isomorphic-ws');
import { Terminal } from 'xterm';
import * as fit from 'xterm/lib/addons/fit/fit';

please use "" for stirng, not single quote, to be same across all of our code. do not use Require, use notation as import * as foo from "Foo"

File: src/Contracts/TerminalService.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    public generateTTYForPod(podName: string, containerName:string):void {
+        const protocols = ['base64url.bearer.authorization.k8s.io.','channel.k8s.io',];
+        //todo: put in a fallback if bin/bash is not found. Should be trying for sh, zsh, csh and powershell for windows containers
+        const path = '/api/v1/namespaces/default/pods/'+podName+'/exec?stdin=1&stdout=1&stderr=1&tty=1&command='+encodeURIComponent("/bin/bash")+'&container='+containerName;
+        const uri = `${this.server}${path}`;
+        //todo: service account already provides token in base64encoded format. Account for the token from service account
+        protocols[0] = protocols[0]+this.token;
+        const sock = new WebSocket(uri, protocols);
+        sock.onopen = () => {
+            this.socket = sock;
```
Comment (madhuv-msft) [inline src/Contracts/TerminalService.ts]: do we want to close if this.socket is not null? or it is opened?

File: src/Contracts/TerminalService.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                });
+                this.terminal.open(this.termRef);
+                (this.terminal as any).fit();
+                this.terminal.focus();
+            }
+        };
+
+        sock.onerror = (err) =>{
+            console.log(err);
+            this.socket = null;
```
Comment (madhuv-msft) [inline src/Contracts/TerminalService.ts]: common code used in more than 2 places

this.socket = null;
            this.terminal && this.terminal.dispose();
            this.terminal = null;

**Comment by madhuv-msft** [comment]
abandoning my review.

but in future when we want to terminal service, we should show the UI in a panel, rather inline.

**Comment by madhuv-msft** [comment]
we will re-open if we have closure on this feature.

**Pull request closed without merging**